File: SortingAlgorithms/SortingAlgorithms/CountingSort.cpp
#include "CountingSort.hpp"

void countingSort(int arr[], int size) {
	int max = arr[0];
	for (int i = 1; i < size; i++) if (max < arr[i]) max = arr[i];

	int* count = new int[max + 1]();
	int* finalArr = new int[size];

	for (int i = 0; i < size; i++) count[arr[i]]++;

	for (int i = 1; i <= max; i++) count[i] += count[i - 1];

	for (int i = size - 1; i >= 0; i--) {
		finalArr[count[arr[i]] - 1] = arr[i];
		count[arr[i]]--;
	}

	for (int i = 0; i < size; i++) arr[i] = finalArr[i];

	delete[] count;
	delete[] finalArr;

}

File: SortingAlgorithmsPython/countingSort.py
def counting_sort(arr):
    max_val = max(arr)

    count = [0] * (max_val + 1)
    final_array = [None] * len(arr)

    for i in range(len(arr)): count[arr[i]] += 1

    for i in range(1, max_val + 1): count[i] += count[i - 1]

    j = len(arr) - 1
    while j >= 0:
        final_array[count[arr[j]] - 1] = arr[j]
        count[arr[j]] -= 1
        j -= 1

    for i in range(len(arr)): arr[i] = final_array[i]